File: n606-construct-string-from-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} t
 * @return {string}
 */
var tree2str = function(t) {
    if (t === null) {
        return ''
    }
    
    let left
    if (t.left !== null || t.right !== null) {
        left = '(' + tree2str(t.left) + ')'
    } else {
        left = ''
    }
    
    let right
    if (t.right !== null) {
        right = '(' + tree2str(t.right) + ')'
    } else {
        right = ''
    }
    
    return t.val + left + right
    
};

File: n606-construct-string-from-binary-tree.go
package main

import "fmt"

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func tree2str(t *TreeNode) string {
	if t == nil {
		return ""
	}

	var left string
	if t.Left != nil || t.Right != nil {
		left = fmt.Sprintf("(%s)", tree2str(t.Left))
	} else {
		left = ""
	}

	var right string
	if t.Right != nil {
		right = fmt.Sprintf("(%s)", tree2str(t.Right))
	} else {
		right = ""
	}

	return fmt.Sprintf("%d%s%s", t.Val, left, right)
}
